src/app/codegen-data-model-provider/CodegenDataModelProvider.cpp

```diff
@@ -396,21 +411,55 @@ bool CodegenDataModelProvider::EndpointExists(EndpointId endpoint)
     return (emberAfIndexFromEndpoint(endpoint) != kEmberInvalidEndpointIndex);
 }
 
-EndpointId CodegenDataModelProvider::FirstEndpoint()
+std::optional<DataModel::EndpointInfo> CodegenDataModelProvider::GetEndpointInfo(EndpointId endpoint)
+{
+    std::optional<unsigned> endpoint_idx = TryFindEndpointIndex(endpoint);
+    if (endpoint_idx.has_value())
+    {
+        return GetEndpointInfoAtIndex(static_cast<uint16_t>(*endpoint_idx));
+    }
+    return std::nullopt;
+}
+
+std::optional<DataModel::EndpointInfo> CodegenDataModelProvider::GetEndpointInfoAtIndex(uint16_t endpointIndex)
+{
+    if (emberAfEndpointIndexIsEnabled(endpointIndex))
```

Comment:
please invert the condition logic to make the conditions as flat as possible. Something like:

```cpp
VerifyOrReturnValue(emberAfEndpointIndexIsEnabled(endpointIndex), std::nullopt);

EndpointId endpointId = emberAfParentEndpointFromIndex(endpointIndex);

if (IsFlatCompositionForEndpointIndex(endpointIndex)) {
    return DataModel::EndpointInfo(endpointId, DataModel::EndpointCompositionPattern::kFullFamilyPattern);
}

if (IsTreeCompositionForEndpointIndex(endpointIndex)) {
  return DataModel::EndpointInfo(endpointId, DataModel::EndpointCompositionPattern::kTreePattern);
}

/// TODO: what if both `Is` methods return false? Is this an error to be logged? the previous code just silently
///              returns a default. If this is the case, why even have both if conditions?

```

Reply:
and followup after reviewing the ember code: what if both would return `true`? I think it would be better if we had an ember API that returns an enum type (likely only full or tree) and has internal assertions that these are mutually exclusive, or if they are not we add more clearly defined enum constants.